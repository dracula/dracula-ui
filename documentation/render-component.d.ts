import React from 'react';
export declare type ComponentExample = {
    html: string;
    docs: string;
    title: string;
    react?: any;
    screenshot?: string;
    svgData?: string;
};
export declare type SnapshotBuilder = {
    comp: () => React.ReactElement;
    docs: string;
    title: string;
};
export declare function snapshot(title: string, comp: () => React.ReactElement, docs?: string): SnapshotBuilder;
export declare function renderSnapshot(title: string, comp: () => React.ReactElement, docs?: string): ComponentExample;
