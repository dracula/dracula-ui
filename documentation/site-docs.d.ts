import { SnapshotBuilder } from './render-component';
import React from 'react';
interface Documentation {
    basic: () => SnapshotBuilder;
    variations: () => Array<SnapshotBuilder> | SnapshotBuilder;
}
export declare function docs<T>(componentClass: React.FunctionComponent<T> | string, documentation: Documentation): void;
export {};
