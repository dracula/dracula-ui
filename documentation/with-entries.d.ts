import { SnapshotBuilder } from './render-component';
export declare function withEntries<T>(map: T, fn: (name: keyof T, value: T[keyof T]) => SnapshotBuilder): SnapshotBuilder[];
