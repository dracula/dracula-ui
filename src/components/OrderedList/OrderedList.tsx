import cx from 'classnames/dedupe'
import React, { OlHTMLAttributes } from 'react'
import { ColorMap } from '../../base/colors'
import {
  cleanProps,
  marginMixin,
  MarginMixin,
  paddingMixin,
  PaddingMixin
} from '../../base/spacing'

export const orderedListColors: Partial<ColorMap> = {
  white: 'drac-list-white',
  cyan: 'drac-list-cyan',
  green: 'drac-list-green',
  orange: 'drac-list-orange',
  pink: 'drac-list-pink',
  purple: 'drac-list-purple',
  red: 'drac-list-red',
  yellow: 'drac-list-yellow'
}

const orderedListTypes = {
  1: 'drac-list-ordered-decimal',
  a: 'drac-list-ordered-lower-alpha',
  A: 'drac-list-ordered-lower-alpha',
  i: 'drac-list-ordered-lower-roman',
  I: 'drac-list-ordered-lower-roman'
}

/**
 * OrderedList Props
 */
export interface OrderedListProps
  extends OlHTMLAttributes<HTMLOListElement>,
    PaddingMixin,
    MarginMixin {
  /** The Dracula UI color for the Ordered List. */
  color?: keyof typeof orderedListColors
  /** The numbering type of the Ordered List, same as a regular `<ol>` element. */
  type?: OlHTMLAttributes<HTMLOListElement>['type']
}

/**
 * Ordered Lists are used to display list items in an ordered way.
 */
export const OrderedList: React.FC<OrderedListProps> = (props) => {
  const { color, type, ...htmlProps } = props

  const finalProps = cleanProps({
    ...htmlProps,
    type,
    className: cx(
      'drac-list drac-list-ordered',
      type && orderedListTypes[type],
      props.className,
      color && orderedListColors[color],
      ...paddingMixin(props),
      ...marginMixin(props)
    )
  })

  return <ol {...finalProps} />
}

OrderedList.displayName = 'OrderedList'
