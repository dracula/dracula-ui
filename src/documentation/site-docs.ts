import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from './render-component'
import React from 'react'
import fs from 'fs-extra'
import { pretty } from './pretty'
import { getDocGen } from '../documentation/docgen/doc-generator'
import { toDSP } from '../documentation/dsp/component-generator'
import puppeteer from 'puppeteer'
import { componentScreenshot } from './screenshots/component-screenshot'
import os from 'os'

interface Documentation {
  basic: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder> | SnapshotBuilder
}

jest.setTimeout(10000)

const RECORDING = process.env.RECORDING

export function docs<T>(
  componentClass: React.FunctionComponent<T> | string,
  documentation: Documentation
) {
  const name =
    typeof componentClass === 'string'
      ? componentClass
      : componentClass.displayName ?? componentClass.name

  const dspPath = `${process.cwd()}/dsp`

  describe(`Site: ${name}`, () => {
    let browser: puppeteer.Browser

    beforeAll(async () => {
      if (!RECORDING) {
        return
      }

      let executablePath = undefined
      if (os.arch() === 'arm64') {
        executablePath = '/Applications/Chromium.app/Contents/MacOS/Chromium'
      }

      browser = await puppeteer.launch({
        headless: true,
        devtools: true,
        executablePath
      })
    })

    let examples: Record<string, ComponentExample> = {}

    afterAll(async () => {
      if (!RECORDING) {
        return
      }

      await browser.close()
      const docGen = getDocGen(name)
      const dsp = toDSP(name, examples, docGen)

      await fs.mkdirp(`${dspPath}/data/components/`).catch(() => {
        // folder already exists
      })

      const path = `${dspPath}/data/components/${name}.json`
      await fs.writeFile(path, pretty(JSON.stringify(dsp), 'json'))
    })

    test('Basic Usage', async () => {
      const variation = documentation.basic()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()

      if (RECORDING) {
        const [screenshot, svg] = await componentScreenshot(
          browser,
          snapshot,
          null,
          name
        )

        snapshot.screenshot = screenshot
        snapshot.svgData = svg
      }

      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations: SnapshotBuilder[] = ([] as SnapshotBuilder[]).concat(
      documentation.variations()
    )

    variations.forEach((variation) => {
      test(variation.title, async () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()

        if (RECORDING) {
          const [screenshot, svg] = await componentScreenshot(
            browser,
            snapshot,
            variation,
            name
          )

          snapshot.screenshot = screenshot
          snapshot.svgData = svg
        }

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
