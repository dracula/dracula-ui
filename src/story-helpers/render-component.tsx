import React from 'react'
import Enzyme, { mount, shallow } from 'enzyme'
import Adapter from '@wojtekmaj/enzyme-adapter-react-17'
import { pretty } from './pretty'

Enzyme.configure({ adapter: new Adapter() })

export type ComponentExample = {
  html: string
  docs: string
  title: string
  react?: any
  screenshot?: string
  svgData?: string
}

export type SnapshotBuilder = {
  comp: () => React.ReactElement
  docs: string
  title: string
}

export function buildSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): SnapshotBuilder {
  return {
    title,
    comp,
    docs: docs.trim()
  }
}

export function renderSnapshot(
  title: string,
  comp: () => React.ReactElement,
  docs: string = ''
): ComponentExample {
  let component = comp()
  const html = mount(component).html()
  let wrapped = false

  // wrap the component in a div when it's not wrapped already
  // we do this so we can get the actual component example to not be shallow rendered.
  // Shallow rendering a parent "div" gives us a nice JSX representation of the example.
  if (component.type !== 'div') {
    component = <div>{component}</div>
    wrapped = true
  }

  let react = shallow(component).debug({ verbose: true })

  if (wrapped) {
    // replace wrapper div ðŸ˜¬
    react = react
      .replace(/^<div>\n/, '')
      .replace(/\n<\/div>$/, '')
      .trim()
  }

  return {
    title,
    html: pretty(html, 'html'),
    react: pretty(react, 'typescript'),
    docs
  }
}
