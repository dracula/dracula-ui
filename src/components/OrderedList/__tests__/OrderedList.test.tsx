import React from 'react'
import { docs } from '@/documentation/site-docs'
import { snapshot } from '@/documentation/render-component'
import { OrderedList } from '@/components/OrderedList/OrderedList'

docs(OrderedList, {
  basic() {
    return snapshot('Usage', Usage, 'Render an ordered list by default.')
  },
  variations() {
    return [
      snapshot(
        'Colors',
        Colors,
        'OrderedLists can be customized to use any of the Dracula UI theme colors.'
      ),
      snapshot(
        'Types',
        Types,
        'OrderedLists can have a numbering type similar to an <ol> element.'
      )
    ]
  }
})

function Usage() {
  return (
    <OrderedList>
      <li className="drac-text drac-text-white">Blade</li>
      <li className="drac-text drac-text-white">Buffy</li>
      <li className="drac-text drac-text-white">Morbius</li>
    </OrderedList>
  )
}

function Colors() {
  return (
    <OrderedList color="cyan">
      <li className="drac-text drac-text-white">Blade</li>
      <li className="drac-text drac-text-white">Buffy</li>
      <li className="drac-text drac-text-white">Morbius</li>
    </OrderedList>
  )
}

function Types() {
  return (
    <OrderedList color="cyan" type="A">
      <li className="drac-text drac-text-white">Blade</li>
      <li className="drac-text drac-text-white">Buffy</li>
      <li className="drac-text drac-text-white">Morbius</li>
    </OrderedList>
  )
}
