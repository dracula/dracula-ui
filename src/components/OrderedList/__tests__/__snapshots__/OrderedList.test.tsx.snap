// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site: OrderedList Basic Usage 1`] = `
Object {
  "docs": "Render an ordered list by default.",
  "html": "<ol class=\\"drac-list drac-list-ordered\\">
  <li class=\\"drac-text drac-text-white\\">Blade</li>
  <li class=\\"drac-text drac-text-white\\">Buffy</li>
  <li class=\\"drac-text drac-text-white\\">Morbius</li>
</ol>
",
  "react": "<OrderedList>
  <li className=\\"drac-text drac-text-white\\">Blade</li>
  <li className=\\"drac-text drac-text-white\\">Buffy</li>
  <li className=\\"drac-text drac-text-white\\">Morbius</li>
</OrderedList>;
",
  "title": "Usage",
}
`;

exports[`Site: OrderedList Colors 1`] = `
Object {
  "docs": "OrderedLists can be customized to use any of the Dracula UI theme colors.",
  "html": "<ol class=\\"drac-list drac-list-ordered drac-list-cyan\\">
  <li class=\\"drac-text drac-text-white\\">Blade</li>
  <li class=\\"drac-text drac-text-white\\">Buffy</li>
  <li class=\\"drac-text drac-text-white\\">Morbius</li>
</ol>
",
  "react": "<OrderedList color=\\"cyan\\">
  <li className=\\"drac-text drac-text-white\\">Blade</li>
  <li className=\\"drac-text drac-text-white\\">Buffy</li>
  <li className=\\"drac-text drac-text-white\\">Morbius</li>
</OrderedList>;
",
  "title": "Colors",
}
`;

exports[`Site: OrderedList Types 1`] = `
Object {
  "docs": "OrderedLists can have a numbering type similar to an <ol> element.",
  "html": "<ol
  type=\\"A\\"
  class=\\"drac-list drac-list-ordered drac-list-ordered-lower-alpha drac-list-cyan\\"
>
  <li class=\\"drac-text drac-text-white\\">Blade</li>
  <li class=\\"drac-text drac-text-white\\">Buffy</li>
  <li class=\\"drac-text drac-text-white\\">Morbius</li>
</ol>
",
  "react": "<OrderedList color=\\"cyan\\" type=\\"A\\">
  <li className=\\"drac-text drac-text-white\\">Blade</li>
  <li className=\\"drac-text drac-text-white\\">Buffy</li>
  <li className=\\"drac-text drac-text-white\\">Morbius</li>
</OrderedList>;
",
  "title": "Types",
}
`;
